File: example/java/jimmer-sql-dynamic/src/main/java/org/babyfish/jimmer/sql/example/bll/interceptor/output/TenantFilterForCacheMode.java
package org.babyfish.jimmer.sql.example.bll.interceptor.output;

import org.babyfish.jimmer.sql.event.EntityEvent;
import org.babyfish.jimmer.sql.example.bll.interceptor.TenantProvider;
import org.babyfish.jimmer.sql.example.model.common.TenantAwareProps;
import org.babyfish.jimmer.sql.filter.CacheableFilter;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

import java.util.SortedMap;
import java.util.TreeMap;

/*
 * see JSqlClient.Builder.addFilters
 *
 * This bean is only be used when cache is used.
 */
@Component
@ConditionalOnProperty("spring.redis.host")
public class TenantFilterForCacheMode
        extends TenantFilterForNonCacheMode
        implements CacheableFilter<TenantAwareProps> {

    public TenantFilterForCacheMode(TenantProvider tenantProvider) {
        super(tenantProvider);
    }

    @Override
    public SortedMap<String, Object> getParameters() {
        String tenant = tenantProvider.get();
        if (tenant == null) {
            return null;
        }
        SortedMap<String, Object> map = new TreeMap<>();
        map.put("tenant", tenant);
        return map;
    }

    @Override
    public boolean isAffectedBy(EntityEvent<?> e) {
        return e.getChangedFieldRef(TenantAwareProps.TENANT) != null;
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/sql/bll/interceptor/output/TenantFilterForCacheMode.kt
package org.babyfish.jimmer.example.kt.sql.bll.interceptor.output

import org.babyfish.jimmer.example.kt.sql.bll.interceptor.TenantProvider
import org.babyfish.jimmer.example.kt.sql.model.common.TenantAware
import org.babyfish.jimmer.sql.event.EntityEvent
import org.babyfish.jimmer.sql.kt.event.getChangedFieldRef
import org.babyfish.jimmer.sql.kt.filter.KCacheableFilter
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Component
import java.util.*

/*
 * see KSqlClientDsl.addFilters
 *
 * This bean is only be used when cache is enabled.
 */
@ConditionalOnProperty("spring.redis.host")
@Component
class TenantFilterForCacheMode(
    tenantProvider: TenantProvider
) : TenantFilterForNonCacheMode(tenantProvider), KCacheableFilter<TenantAware> {

    override fun getParameters(): SortedMap<String, Any>? =
        tenantProvider.tenant?.let {
            sortedMapOf("tenant" to it)
        }

    override fun isAffectedBy(e: EntityEvent<*>): Boolean =
        e.getChangedFieldRef(TenantAware::tenant) !== null
}